impeller/renderer/backend/vulkan/swapchain/ahb/ahb_swapchain_impl_vk.cc

```diff
@@ -208,17 +221,140 @@ std::shared_ptr<ExternalFenceVK> AHBSwapchainImplVK::SubmitCompletionSignal(
   return fence;
 }
 
-void AHBSwapchainImplVK::OnTextureSetOnSurfaceControl(
+vk::UniqueSemaphore AHBSwapchainImplVK::CreateRenderReadySemaphore(
+    const std::shared_ptr<fml::UniqueFD>& fd) const {
+  if (!fd->is_valid()) {
+    return {};
+  }
+
+  auto context = transients_->GetContext().lock();
+  if (!context) {
+    return {};
+  }
+
+  const auto& context_vk = ContextVK::Cast(*context);
+
+  const auto& device = context_vk.GetDevice();
+
+  auto signal_wait = device.createSemaphoreUnique({});
+
+  if (signal_wait.result != vk::Result::eSuccess) {
+    return {};
+  }
+
+  context_vk.SetDebugName(*signal_wait.value, "AHBRenderReadySemaphore");
+
+  vk::ImportSemaphoreFdInfoKHR import_info;
+  import_info.semaphore = *signal_wait.value;
+  import_info.fd = fd->get();
+  import_info.handleType = vk::ExternalSemaphoreHandleTypeFlagBits::eSyncFd;
+  // From the spec: Sync FDs can only be imported temporarily.
+  import_info.flags = vk::SemaphoreImportFlagBitsKHR::eTemporary;
+
+  const auto import_result = device.importSemaphoreFdKHR(import_info);
+
+  if (import_result != vk::Result::eSuccess) {
+    VALIDATION_LOG << "Could not import semaphore FD: "
+                   << vk::to_string(import_result);
+    return {};
+  }
+
+  // From the spec: Importing a semaphore payload from a file descriptor
+  // transfers ownership of the file descriptor from the application to the
+  // Vulkan implementation. The application must not perform any operations on
+  // the file descriptor after a successful import.
+  [[maybe_unused]] auto released = fd->release();
+
+  return std::move(signal_wait.value);
+}
+
+bool AHBSwapchainImplVK::SubmitWaitForRenderReady(
+    const std::shared_ptr<fml::UniqueFD>& render_ready_fence,
+    const std::shared_ptr<AHBTextureSourceVK>& texture) const {
+  // If there is no render ready fence, we are already ready to render into
+  // the texture. There is nothing more to do.
+  if (!render_ready_fence || !render_ready_fence->is_valid()) {
+    return true;
+  }
+
+  auto context = transients_->GetContext().lock();
+  if (!context) {
+    return false;
+  }
+
+  auto command_buffer = context->CreateCommandBuffer();
+  if (!command_buffer) {
+    return false;
+  }
+  command_buffer->SetLabel("AHBSubmitWaitForRenderReadyCB");
+  const auto& encoder = CommandBufferVK::Cast(*command_buffer).GetEncoder();
+
+  const auto command_encoder_vk = encoder->GetCommandBuffer();
+
+  BarrierVK barrier;
+  barrier.cmd_buffer = command_encoder_vk;
+  barrier.new_layout = vk::ImageLayout::eColorAttachmentOptimal;
+  barrier.src_stage = vk::PipelineStageFlagBits::eBottomOfPipe;
+  barrier.src_access = {};
+  barrier.dst_stage = vk::PipelineStageFlagBits::eTopOfPipe;
+  barrier.dst_access = {};
+
+  if (!texture->SetLayout(barrier).ok()) {
+    return false;
+  }
+
+  auto render_ready_semaphore =
+      MakeSharedVK(CreateRenderReadySemaphore(render_ready_fence));
+
+  encoder->Track(render_ready_semaphore);
```

Comment:
The fix for the early destruction is tracking the semaphore then?

Reply:
My intuition is that this should have done the trick. But the semaphore is being dropped too early and I'm trying to debug why.

Reply:
From the discussion on Discord. I am misusing the resource tracker because I am not creating passes and such and directly dealing with VK objects. I need to hold the encoder alive for long enough till a completion fence (which doesn't exist yet) is waited upon by the fence waiter.